src/adapter.rs

```diff
@@ -0,0 +1,461 @@
+#[cfg(all(target_os = "linux", feature = "bluetooth"))]
+use blurz::bluetooth_adapter::BluetoothAdapter as BluetoothAdapterBluez;
+#[cfg(all(target_os = "android", feature = "bluetooth"))]
+use blurdroid::bluetooth_adapter::Adapter as BluetoothAdapterAndroid;
+#[cfg(all(target_os = "macos", feature = "bluetooth"))]
+use blurmac::BluetoothAdapter as BluetoothAdapterMac;
+#[cfg(not(any(all(target_os = "linux", feature = "bluetooth"),
+              all(target_os = "android", feature = "bluetooth"),
+              all(target_os = "macos", feature = "bluetooth"))))]
+use empty::BluetoothAdapter as BluetoothAdapterEmpty;
+#[cfg(feature = "bluetooth-test")]
+use blurmock::fake_adapter::FakeBluetoothAdapter;
+
+#[cfg(feature = "bluetooth-test")]
+const NOT_SUPPORTED_ON_REAL_ERROR: &'static str = "Error! Test functions are not supported on real devices!";
+#[cfg(feature = "bluetooth-test")]
+const NOT_SUPPORTED_ON_MOCK_ERROR: &'static str = "Error! The first parameter must be a mock structure!";
+
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::error::Error;
+use bluetooth::BluetoothDiscoverySession;
+use bluetooth::BluetoothDevice;
+
+pub trait BluetoothAdapter {
+    fn new()-> Result<Box<BluetoothAdapter>, Box<Error>>;
+    fn get_id(&self)-> String;
+    fn get_devices(&self)-> Result<Vec<BluetoothDevice>, Box<Error>>;
+    fn get_device(&self, address: String) -> Result<Option<BluetoothDevice>, Box<Error>>;
+    fn get_address(&self) -> Result<String, Box<Error>>;
+    fn get_name(&self)-> Result<String, Box<Error>>;
+    fn get_alias(&self) -> Result<String, Box<Error>>;
+    fn get_class(&self)-> Result<u32, Box<Error>>;
+    fn is_powered(&self)-> Result<bool, Box<Error>>;
+    fn is_discoverable(&self) -> Result<bool, Box<Error>>;
+    fn is_pairable(&self)-> Result<bool, Box<Error>>;
+    fn get_pairable_timeout(&self) -> Result<u32, Box<Error>>;
+    fn get_discoverable_timeout(&self)-> Result<u32, Box<Error>>;
+    fn is_discovering(&self)-> Result<bool, Box<Error>>;
+    fn create_discovery_session(&self) -> Result<BluetoothDiscoverySession, Box<Error>> ;
+    fn get_uuids(&self)-> Result<Vec<String>, Box<Error>>;
+    fn get_vendor_id_source(&self)-> Result<String, Box<Error>>;
+    fn get_vendor_id(&self)-> Result<u32, Box<Error>>;
+    fn get_product_id(&self) -> Result<u32, Box<Error>> ;
+    fn get_device_id(&self) -> Result<u32, Box<Error>>;
+    fn get_modalias(&self) -> Result<(String, u32, u32, u32), Box<Error>>;  
+}
+pub struct Bluez(Arc<BluetoothAdapterBluez>);
```

Comment:
Each of the per-platform implementations should not be public, and will need a `#cfg[...]` block to ensure they are only visible to the compiler on appropriate platforms.

Reply:
May I also add a suggestion? I see that there's a lot of methods on this struct which basically delegates back to `self.0.method()`, so is it possible to instead implement Deref<Target = BluetoothAdapterBluez> instead of delegating methods?

Reply:
That exposes the per-platform types to the caller crate. I prefer to have the trait as a layer of indirection to prevent that, and expose any cross-platform API issues.